### ISSUE TITLE
tensor.to(device) not copying data correctly between two GPUs

### ISSUE THREAD

**Issue by nishitvp** [pytorch/pytorch#132397]
### 🐛 Describe the bug

It seems that tensor.to(device) isn't copying the data correctly when it is moved from one GPU to another GPU.
Copying to/fro a GPU and a CPU seems to work as intended.

Minimal example -
```
import torch

# Tensor initialized on GPU
test_tensor = torch.tensor([1, 2, 3], dtype=torch.float16, device=0)

# Moving between two GPUs, tensor becomes [0, 0, 0] instead of [1, 2, 3] on device="cuda:1"
new_tensor = test_tensor.to(1)

# Moving from GPU to CPU, works as expected with the correct data - [1, 2, 3] on the correct device
new_2_tensor = test_tensor.to("cpu")

# Moving from CPU to GPU, works as expected with the correct data - [1, 2, 3] with device="cuda:2"
new_3_tensor = new_2_tensor.to(2)

# Moving between two GPUs - becomes [0, 0, 0] instead of [1, 2, 3] with device = "cuda:3"
new_4_tensor = new_3_tensor.to(3)
```

Machine specs -
AWS g6.48xlarge dedicated host with 8 L4 GPUs on it.
Pytorch version - 2.4.0

### Versions

PyTorch version: 2.4.0
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04 LTS (x86_64)
GCC version: (Ubuntu 13.2.0-23ubuntu4) 13.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-1012-aws-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA L4
GPU 1: NVIDIA L4
GPU 2: NVIDIA L4
GPU 3: NVIDIA L4
GPU 4: NVIDIA L4
GPU 5: NVIDIA L4
GPU 6: NVIDIA L4
GPU 7: NVIDIA L4

Nvidia driver version: 555.42.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               192
On-line CPU(s) list:                  0-191
Vendor ID:                            AuthenticAMD
Model name:                           AMD EPYC 7R13 Processor
CPU family:                           25
Model:                                1
Thread(s) per core:                   2
Core(s) per socket:                   48
Socket(s):                            2
Stepping:                             1
BogoMIPS:                             5299.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save vaes vpclmulqdq rdpid
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            3 MiB (96 instances)
L1i cache:                            3 MiB (96 instances)
L2 cache:                             48 MiB (96 instances)
L3 cache:                             384 MiB (12 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-47,96-143
NUMA node1 CPU(s):                    48-95,144-191
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.4.0
[pip3] torchaudio==2.4.0
[pip3] torchvision==0.19.0
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.8           py310h5eee18b_0  
[conda] mkl_random                1.2.4           py310hdb19cb5_0  
[conda] numpy                     1.26.4          py310h5f9d8c6_0  
[conda] numpy-base                1.26.4          py310hb5e798b_0  
[conda] pytorch                   2.4.0           py3.10_cuda12.4_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              12.4                 hc786d27_6    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.4.0               py310_cu124    pytorch
[conda] torchtriton               3.0.0                     py310    pytorch
[conda] torchvision               0.19.0              py310_cu124    pytorch

**Comment by malfet** [comment]
High pri to get a repro (which seems to be deterministic)
Abovementioned code works for me on `p4dn.24xlarge`

**Comment by nishitvp** [comment]
UPDATE: I could also reproduce the same bug on a completely different dedicated host of the same type (AWS g6.48xlarge with 8x L4s).

**Comment by ptrblck** [comment]
Could you check if disabling IOMMU fixes the issue as described [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html#pci-access-control-services-acs)? It would also be great if you could run a pure p2p test (without using PyTorch) to check if this is a system issue and unrelated to PyTorch.

**Comment by nishitvp** [comment]
I ran the test at https://github.com/NVIDIA/cuda-samples/blob/master/Samples/5_Domain_Specific/p2pBandwidthLatencyTest/p2pBandwidthLatencyTest.cu

Seemingly, all devices can access each other, but the latency seems to increase massively (and bandwidth decreases massively) when P2P is enabled (?!?!?!)

Here are the logs of the test - 

[p2p_test.log.txt](https://github.com/user-attachments/files/16461156/p2p_test.log.txt)

I hope I am not misinterpreting the results. I suspect this is a lower-level system issue?

**Comment by ezyang** [comment]
Downgrading high prio as it's likely an issue with CUDA/the hardware setup

**Comment by nishitvp** [comment]
Indeed, the issue went away when I switched from Ubuntu to Amazon Linux, which indicates that it might be an OS/CUDA interface issue.

**Comment by msaroufim** [comment]
I couldn't repro on CentOS but please feel free to reopen if you run into this again
